Add eslint support

sam-github opened the pull request:
I gave up on line lengths and eqeqeq support, probes in particular would need to substantially rewritten to not have run-on lines.

There are a couple dozen errors remaining and which appear to me to be genuine bugs (eslint isn't all style).

I recommend a close review while ignoring whitespace: https://github.com/RuntimeTools/appmetrics/pull/274/files?w=1

I don't know if tests still pass, see #271

gibfahn commented:
@sam-github could you rebase/update this PR? It'd be good to get some linting in.

gibfahn commented:
ping @sam-github

sam-github commented:
@gibfahn It didn't get landed the first time, so without a proactive confirmation that @tobespc is interested in merging it once updated, I won't spend more time on it.

tobespc commented:
@sam-github Apologies Sam for leaving this so long.  Yes, I would like eslint to go in.  The appmetrics codebase is now fixed ready for the next release so could you rebase and merge and I will accept the change in before we ship

sam-github commented:
@tobespc and all the rest of you, PTAL, this time I'd like to merge this *before* the change rots.

Note that there are outstanding lint violations that I did not fix because they are errors in the code that I'm not sure how to fix. I'll comment on my own PR to highlight them (they all have `FIXME(sam) ...` in code).

Alternatively, you can just run `npm pretest` on this branch and the failures show you were the errors are.

```
% npm run pretest

> appmetrics@3.0.2 pretest /home/sam/w/sn/appmetrics
> eslint .


/home/sam/w/sn/appmetrics/headless_zip.js
  111:5  error  'ouputDir' is not defined  no-undef

/home/sam/w/sn/appmetrics/lib/request.js
  73:9  error  'parentId' is not defined                      no-undef
  75:9  error  'parentId' is not defined                      no-undef
  95:9  error  'parentId' is assigned a value but never used  no-unused-vars

/home/sam/w/sn/appmetrics/probes/leveldown-probe.js
  133:3   error  'req' is not defined  no-undef
  134:3   error  'req' is not defined  no-undef
  138:35  error  'req' is not defined  no-undef

/home/sam/w/sn/appmetrics/probes/mqlight-probe.js
  65:57  error  'methodArgs' is not defined  no-undef
  70:15  error  'methodArgs' is not defined  no-undef

✖ 9 problems (9 errors, 0 warnings)
```

sam-github commented on headless_zip.js:
```diff
@@ -13,98 +13,112 @@
+    // FIXME(sam) not sure what to do here, this code probably never did
+    // anything but I don't want to enable it in error (btw,  cwd() always
+    // returns a string).
+    ouputDir = process.cwd().toString();
```
See comment above. The code could be rewritten as `var outputDir = dirToWriteTo || process.cwd();`, but since code has never actually done anything, perhaps it should just be deleted?

sjanuary replied:
Looks like just a missing 't' in ou**t**putDir line 96 (or 111)?  This code is supposed to create a headless archive on process exit if running in headless mode. I did test it when I wrote it but I guess I must have tested it with the other scenario (i.e. manually specifying the output directory) and missed this one.

sam-github replied:
Yes, its a missing `t`, but if I just blindly correct the spelling, the code will suddenly do something it didn't used to do. I'll do it if you confirm its correct. I don't know what headless mode is, or why the string returned from process.cwd() is being turned into another string.

sjanuary replied:
Yes, please add the 't' and `toString()` can be removed if it's superfluous.  Or I can do all the fixes under a separate PR if you would prefer?

sam-github replied:
done

sam-github commented on lib/request.js:
```diff
@@ -32,175 +33,192 @@ var am = require('../');
+    if (this.parent) {
+      this.parent.traceStart();
+      // FIXME(sam) assigning to a global parentId that appears to never be
+      // used, but I can't be sure.
```
Just deleting this section of code which appears to do nothing makes me nervous, its possible a global `parentId` is being used to track requests... except that would almost certainly be a bug.

sjanuary replied:
`parentId` doesn't seem to be used anywhere in appmetrics, either in requests.js or out of it so I think it should be safe to delete both occurrences.

sam-github replied:
done

sam-github commented on lib/request.js:
```diff
@@ -32,175 +33,192 @@ var am = require('../');
+    // this whole chunk of code can be deleted, or the `var parentId` is a bug,
+    // and this is supposed to be setting a global, except that wouldn't be
+    // parallel request safe... not sure what to do here.
+    var parentId;
```
Here as well, its not assigning to a global, so the entire section is a no-op... which makes me think that it is supposed to be assigning to a global, and that the `var` is a bug... except using a global would be a bug... So, over to you.

sjanuary replied:
As above, I think this is just an unused variable that could be deleted

sam-github replied:
done

sam-github commented on probes/leveldown-probe.js:
```diff
@@ -13,97 +13,129 @@
+  // FIXME(sam) req is used as a global to communicate with requestEnd, almost
+  // certainly a bug, what happens if two requests are started before the first
+  // ends?
+  req = request.startRequest('leveldown', 'query');
```
This looks like a bug that would damage the tracking of simultaneous parallel requests.

sjanuary replied:
Other probes use `probeData.req`, I think this should do the same.  Again probably untested.  Argh!

sam-github replied:
so, what to do? If I merge this branch as-is, it will start to fail all tests. I think I'll leave the FIXME, but add a eslint inline config statement to ignore the lint error, and then you can PR master to fix it as you get a chance.

sam-github replied:
I disabled eslint for these lines, but I'm pretty sure from the lines a bit futher down:

```
if (probeData && probeData.req) req.stop(...
```

That just putting `probeData.` before all the `req` would make this work, but since I don't have the setup to test this change, I'll leave that bug fix to you @sjanuary

sam-github commented on probes/mqlight-probe.js:
```diff
@@ -13,166 +13,239 @@
-                    aspect.strongTraceTransactionLink('mqlight: ', methodName, methodArgs[callbackPosition]);
-                }
+          // Call the transaction link with a name and the callback for strong trace
+          var callbackPosition = aspect.findCallbackArg(methodArgs); // FIXME(sam) methodArgs does not exist!
```
I suspect this is supposed to be `args` from line 52, but I don't know the code so can't be sure. Also, since it can't be working now, it is probably untested.

sjanuary replied:
This probe was tested when originally written.  I suspect https://github.com/RuntimeTools/appmetrics/pull/257 was never tested since it seems to span all the probes...  Slightly concerning...  Anyway looking at the other changes in that PR it seems like it should be `args`.

sam-github replied:
should I change it, or eslint-ignore the warning and let you change it and add tests?

sjanuary replied:
I think just change it if you're happy to do that, or I can do it as a separate PR.  We don't really have the cycles right now to be improving the probes tests unfortunately, we just rely on manual testing by the developer at the time of writing which obviously didn't happen in this case...

sam-github replied:
I don't know how to manually test, but I made the changes.

sjanuary commented on extract_all_binaries.js:
```diff
@@ -25,149 +25,189 @@ var OS = process.platform; // e.g. linux
+  console.log(
+    '********************************************************************************'
+  );
 };
```
Does eslint add an 80 character line limit?  Do we have to have that?  Personally I think changes like this actually decrease the readabililty and 80 characters is pretty small for modern monitors.  Also this PR introduces a lot of new warnings from codebeat about function lengths being too long, which I guess are related to adding a line limit.

sam-github replied:
not sure what `codebeat` is, but if its algorithm allows joining lines together to make functions "shorter and more readable", I think its algorithm has problems! I'll change the limit to 120, though.

sam-github replied:
Also, the translation to fit line length was mechanical, a human would write the above as `console.log("*".repeat(80))`, or whatever the length is - I can't actually see the end of the line because I'm reading this in github, and while people's monitors and personal editor setup may be huge, the github UI width seems to be fixed, so the right side of the line is truncated.

sjanuary replied:
It's annoying that github restricts it to 80, I've got masses of white space either side in my browser.  I think the overall preference in this team is for a width that works well in an editor/IDE although I can see why some people like to stick to 80. I agree that codebeat may not be that useful though...

sam-github replied:
Btw, note that on >- 4.x, node supports
```
console.log(`
this is a long... imagine it is anyhow
this is a paragraph
so doesn't need multiple console.log
`);
```

gibfahn replied:
@sam-github but that is a bit annoying when you're indented:

```js
      console.log(`
this is a long... imagine it is anyhow
this is a paragraph
so doesn't need multiple console.log
You can't have whitespace at the beginning of the line
or it will be included in your string.
      `);
```

sam-github replied:
that's better then having every line prefixed with console.log. And it gets even better when you change the text and have to rewrap the lines. Butal when its a bunch of lines like `'one line' +`, but any text editor can do it when it uses backticks. And anyhow, this would normally be written like:
```
var licenseWarning = `
blah
blah
`;

function whatever() {
   console.log(licenseWarning)
}
```

tobespc commented:
@sjanuary as discussed on st, I agree with you that 100 / 120 would make more sense

gibfahn commented:
+1 to 120 as the limit, it works well on my screens.

sam-github commented:
eslint passes locally, should be good to merge, @sjanuary PTAL

Btw, I thought  this test had failed, but it was just suspiciously slow:

```
    # Subtest: Headless mode should produce a .hcd file

        ok 1 - nodeappmetrics190617_114329_21314.hcd HCD file found
        1..1
    ok 1 - Headless mode should produce a .hcd file # time=70022.173ms
    # time=70031.922ms
```

70 seconds is a long time, is it expected?

tobespc commented:
@sam-github 70 seconds is a long time but depending on what else the machine is doing its not unheard of.   The creation of the hcd is basically a zip operation taking place so factors such as file size, other disk activity taking place etc can all impact the time

sjanuary commented:
@sam-github it's because the headless settings (com.ibm.diagnostics.healthcenter.headless.run.duration property) takes a number as a parameter which is the run duration in minutes.  So minimum run duration is 1 minute.

sjanuary commented on .eslintrc:
```diff
@@ -6,7 +6,7 @@
   "rules": {
     "eqeqeq": "off",
-    "max-len": "off",
+    "max-len": ["error", 100, 8, {"ignoreComments": true}],
```
I think 120 was the general preference when we talked about this on Slack

sam-github replied:
will change

sam-github commented:
> the run duration in minutes

sneek in support for 0.1 minutes and it would go faster! ;-)

Updated the line length, I think I've addressed all comments and this is ready to merge, PTAL

sam-github commented:
Also re-ran `prettier` (`npm run fix-prettier`). Its an opionated code-reformatter, I recommend it for any code you have to at least start it off on a consistent footing.

@tobespc @sjanuary @stalleyj @hhellyer If you have any open PRs that don't lint after this merges, fix them like so: `npm run fix-prettier; npm run fix-eslint`, anything left over is probably a non-stylistic problem.

sam-github commented:
Its possible that the introduction  of consistent formatting has allowed codebeat to notice cut-n-pasted code in the probes, its got a number of "introduced" issues related to that. Kindof sucky that the issues it finds don't have URLs that link directly to them.

sjanuary commented:
Now merged.  Thanks @sam-github for all the work that went in to this.
